### PR TITLE
Move moderation links into operations collection.

### PR DESCRIPTION
**Pull request by johangant**
Use query parameter to toggle message confirmation from the controller

### CODE REVIEW

File: origins_workflow/origins_workflow.module [hunk 1]
```diff
@@ -448,7 +450,77 @@ function origins_workflow_preprocess_page(&$variables) {
...
+  // Use the core service to fetch the permissable transitions based
+  // on underlying configuration and active user object.
+  $transitions = \Drupal::service('content_moderation.state_transition_validation')
+    ->getValidTransitions($entity, \Drupal::currentUser());
+
+  // Remove the current moderation status from the list of possible transitions.
+  // Quick publish isn't detected as an equivalent operation/status as published
+  // so we prune it off here for now as it makes no sense ot quick publish an
+  // already published piece of content.
+  if ($entity->moderation_state->value === 'published') {
```
Comment (johangant) [inline origins_workflow/origins_workflow.module:477]: There are some semantic quirks here that likely either need some custom weeding/filtering. The service ought to be doing this but I think our slight adjustment on core's default workflow means we have some extra work to do here. Essentially, we just want to weed out the options that make no sense/are duplicates for each given moderation state. Eg: no publish option for something that's already published.

File: origins_workflow/origins_workflow.module [hunk 2]
```diff
@@ -448,7 +450,77 @@ function origins_workflow_preprocess_page(&$variables) {
...
+
+  // We're going to use the permitted transitions returned above
+  // to create a collection of Url objects which we'll use to inject
+  // into the links collection. That allows editors to then select specific
+  // tasks to occur from a single location.
+
+  // Magic numbers... scheduled_transitions.module starts at 50 so a few
+  // between that and 0 (we want the 'Edit' link first) is fine.
+  $weight = 40;
+  foreach ($transitions as $id => $transition) {
```
Comment (johangant) [inline origins_workflow/origins_workflow.module:489]: These Url objects replace what was printed by views in different table columns before.

File: origins_workflow/origins_workflow.module [hunk 3]
```diff
@@ -448,7 +450,77 @@ function origins_workflow_preprocess_page(&$variables) {
...
+
+    $url = Url::fromRoute('origins_workflow.moderation_state_controller_change_state', [
+      'nid' => $entity->id(),
+      'new_state' => $transition->to()->id(),
+    ],
+    [
+      'query' => [
+        'destination' => \Drupal::destination()->get(),
+        // Confirm query parameter triggers a message back to the user from the controller.
+        'confirm' => TRUE,
```
Comment (johangant) [inline origins_workflow/origins_workflow.module:504]: There's no form POST so no way to trigger the message update, so I cheated and sneaked in a query parameter which the controller can detect and set a message there.

File: origins_workflow/origins_workflow.module [hunk 4]
```diff
@@ -448,7 +450,77 @@ function origins_workflow_preprocess_page(&$variables) {
...
+      'title' => \t('Change to @label', ['@label' => $transition->to()->label()]),
+      'url' => $url,
+      'weight' => $weight,
+    ];
+
+    $weight--;
+  }
+
+  // Add an operation to view the moderation history of this content.
+  $operations['view_moderation_history'] = [
```
Comment (johangant) [inline origins_workflow/origins_workflow.module:518]: Another link to remove more views rewrite output config. Replicates the sort of thing you see in the moderation sidebar 👍

File: origins_workflow/src/Controller/ModerationStateController.php [hunk 5]
```diff
@@ -54,8 +69,16 @@ public static function create(ContainerInterface $container) {
    */
   public function changeState($nid, $new_state) {
     // Load the entity.
+    /** @var \Drupal\node\NodeInterface $entity */
     $entity = $this->entityTypeManager->getStorage('node')->load($nid);
-    if ($entity) {
+
+    /** @var \Drupal\workflows\StateInterface $new_state_entity */
+    $new_state_entity = $this->moderationInformation
```
Comment (johangant) [inline origins_workflow/src/Controller/ModerationStateController.php:76]: Hilariously convoluted! Took me ages to get this working.

File: origins_workflow/src/Controller/ModerationStateController.php [hunk 6]
```diff
@@ -69,6 +92,15 @@ public function changeState($nid, $new_state) {
           '@user' => $this->currentUser()->getAccountName(),
         ]);
         $this->logger->notice($message);
+
+        if (!empty(\Drupal::request()->query->get('confirm'))) {
+          $message = t('Moderation state of "@title" changed to @new_state', [
+            '@title' => $entity->getTitle(),
+            '@new_state' => $new_state_entity->label(),
```
Comment (johangant) [inline origins_workflow/src/Controller/ModerationStateController.php:99]: Getting the moderation state object to allow me to pull the label out for it required some mental gymnastic here. You essentially need to use a service to pull the plugin class which in turn allows you to load the state object. Phew

File: origins_workflow/src/Controller/ModerationStateController.php [hunk 7]
```diff
@@ -69,6 +92,15 @@ public function changeState($nid, $new_state) {
...
+          ]);
+
+          \Drupal::messenger()->addMessage($message, \Drupal::messenger()::TYPE_STATUS);
```
Comment (johangant) [inline origins_workflow/src/Controller/ModerationStateController.php:102]: Ah, this needs to be passed in by dependency injection. I'll sort that before merging!

Comment (johangant) [reply origins_workflow/src/Controller/ModerationStateController.php]: Added services for messenger and also request stack

File: origins_workflow/origins_workflow.module [hunk 8]
```diff
@@ -448,7 +450,77 @@ function origins_workflow_preprocess_page(&$variables) {
...
+    'title' => \t('View moderation history'),
+    'url' => Url::fromRoute('entity.node.version_history', ['node' => $entity->id()]),
```
Comment (omahm) [inline origins_workflow/origins_workflow.module:520]: With the first check in this hook being 'ContentEntityInterface' I think you're going to catch non-node entities such as taxonomy terms, GP's etc in this part and try and assign a node route.

**Review by omahm — COMMENTED** [review]
Good stuff and excellent comments!
I've one wee query regarding the creation of node links against matches to ContentEntityInterface

File: origins_workflow/origins_workflow.module (hunk 8 above)
Comment (johangant) [reply origins_workflow/origins_workflow.module:520]: Good spot! I'll double check this, not quite sure why I left it so generalised but probably no harm to drill down to test for NodeInterface as that's what we're dealing with

**Review by omahm — APPROVED** [review]
LG

**Review by DuttonMa — APPROVED** [review]
Looks great, like the comments !

**Pull request merged**